### ISSUE TITLE
[BUG] Drag and drop from Google Chrome to Firefox fails

### ISSUE THREAD

**Issue by pearson** [Jermolene/TiddlyWiki5#7440]
### Describe the bug

Dragging and dropping a plugin link from Google Chrome to Firefox on a Windows 11 PC fails to correctly import the plugin.

This problem only occurs when dragging and dropping a plugin from Google Chrome to a wiki in Firefox. It does not happen when dragging and dropping a plugin "in the opposite direction" from Firefox to a wiki Google Chrome, or when dragging and dropping between two windows of the same browser.

Edit: As might be expected, this problem also happens when dragging and dropping a plugin from Microsoft's Edge to a wiki in Firefox. It also works correctly when dragging and dropping a plugin from Firefox to a wiki in Edge.

### Expected behavior

I would expect the plugin to be correctly imported.

### To Reproduce

1. In Google Chrome, open a plugin page such as [Section Editor](https://kookma.github.io/TW-Section/#%24%3A%2Fplugins%2Fkookma%2Fsection)
2. In Firefox, open a Tiddlywiki wiki, such as `empty.html` from [this link](https://tiddlywiki.com/#GettingStarted)
3. Drag the plugin link from the Google Chrome page to the wiki in Firefox, and drop the link once the "Drop now" banner appears
4. Notice that the import title is "Untitled" and not "$:/plugins/kookma/section"
5. Click the down arrow next to "Untitled" and notice that there is no code shown: ![chrome to firefox drag and drop example - with down arrow clicked](https://user-images.githubusercontent.com/656880/236949683-1deda94f-0977-44e8-a3ed-e913b58578db.png)
6. Click "Import" and notice that no plugin was imported. (Check Control Panel -> Settings -> Plugins, for example.)

### Screenshots

The expected result of this drag and drop is shown in this screenshot:

![firefox to firefox drag and drop example - with down arrow clicked](https://user-images.githubusercontent.com/656880/236949800-2590d52c-1d7e-43af-bbe0-a32895216c08.png)

### TiddlyWiki Configuration

- The default `empty.html` v5.2.7
- Saving mechanism: nothing changed from the default `empty.html`
- Plugins installed: nothing changed from the default `empty.html`

### Desktop (please complete the following information):

- OS: Windows 11
- Google Chrome v112.0.5615.138 64-bit 
- Firefox v112.0.2 64-bit

**Comment by Jermolene** [comment]
Thanks @pmario. It seems that we need to do a careful review of the drag/drop/clipboard processing. It would be great if you could use "git blame" to track the source of the problem.

**Comment by saqimtiaz** [comment]
My guess would be that the responsible PR is this one: https://github.com/Jermolene/TiddlyWiki5/pull/6789 however I do not have the opportunity to test and debug at present.

Whilst this PR was tested across different browsers, that did not include testing drag and drop between different browser combinations, which quite frankly would be a ludicrous amount of work and means we should be even more careful when making changes that impact drag and drop.

If the PR mentioned is indeed the culprit, reverting it will break drag and drop on Android as explained in #6617 in which case I would appreciate any input on how we might address that as well.

**Comment by pmario** [comment]
>If the PR mentioned is indeed the culprit, reverting it will break drag and drop on Android as explained in https://github.com/Jermolene/TiddlyWiki5/issues/6617 in which case I would appreciate any input on how we might address that as well.

Yea d&d is "sneaky" ... I'll have a closer look at the PR Saq mentioned later today

**Comment by pmario** [comment]
related to: **test-with-mobile-chrome** #7445

**Comment by pmario** [comment]
TLDR;

I did try to iron out the problem without browser detection and I failed. The differences in browser implementations of the drag & drop mechanism and the `dataTransfer` object are so subtle, that fixing it for one browser breaks one or all other browsers. 

------

@Jermolene, @saqimtiaz ... We need to fix this problem, that got introduced with https://github.com/Jermolene/TiddlyWiki5/pull/6789 ... 

The fix at #6789 used the IE browser detection to move some code out of the default program path into an IE only code path. That causes several issues including this one. 

- It would have been needed to identify Chrome or mobile chrome and create a new working path for chrome. 
- The existing default path should have been left alone. 

I know, that we should use feature detection to solve problems like this one, but that's not really possible, because all browsers involved will report, that the feature is there. .. BUT the implementation details are different in different browsers. 

We do have the [browser-sniff plugin](https://tiddlywiki.com/#%24%3A%2Fplugins%2Ftiddlywiki%2Fbrowser-sniff), which IMO would make it possible to tackle this problem. But the browser detection functions are not part of the core. 

**So my proposal would be to add the browser detection to the core and use it there to fix the problem.** 

What do you think?

**Comment by pmario** [comment]
@pearson Can you test it with this preview https://tiddlywiki5-git-fork-pmario-fix-dnd-for-mobile-touch-jermolene.vercel.app/ from PR https://github.com/Jermolene/TiddlyWiki5/pull/7622 ?

**Comment by Jermolene** [comment]
> So my proposal would be to add the browser detection to the core and use it there to fix the problem.

Hi @pmario the browser-sniff plugin uses a third party library which is perhaps excessive for the core. But adding more $tw.browser.* properties alongside the existing ones seems reasonable.

**Comment by pmario** [comment]
@pearson ... There has been a PR merged today. So please check with: https://tiddlywiki.com/prerelease/  and report back. 

If the problem is solved please close the issue.

**Comment by pmario** [comment]
@pearson This one should be fixed with TW v5.3.1 now. So you should close it

**Comment by pmario** [comment]
@Jermolene -- Please close since fixes are merged.

**Comment by Jermolene** [comment]
Thank you @pmario
